### PR TITLE
benchmark with float16

### PR DESCRIPTION
**Pull request by zheng-da**
## Description ##
(Brief description on what this PR is about)

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA issue](https://issues.apache.org/jira/projects/MXNET/issues) created (except PRs with tiny changes)
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [ ] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [ ] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### CODE REVIEW

File: example/image-classification/benchmark_score.py [hunk 1]
```diff
@@ -74,11 +85,15 @@ def score(network, dev, batch_size, num_batches):
     devs.append(mx.cpu())
 
     batch_sizes = [1, 2, 4, 8, 16, 32]
+    dtype='float32'
 
     for net in networks:
+        # AlexNet doesn't support float16
+        if (net == 'alexnet' and dtype == 'float16'):
+            continue
```
Comment (marcoabreu) [inline example/image-classification/benchmark_score.py]: Please print a message

**Comment by szha** [comment]
ping

File: example/image-classification/benchmark_score.py [hunk 2]
```diff
@@ -39,21 +39,32 @@ def get_symbol(network, batch_size):
...
              inputs_need_grad = False,
              data_shapes      = data_shape)
     mod.init_params(initializer=mx.init.Xavier(magnitude=2.))
 
     # get data
-    data = [mx.random.uniform(-1.0, 1.0, shape=shape, ctx=dev) for _, shape in mod.data_shapes]
+    data = [mx.random.uniform(-1.0, 1.0, shape=shape, ctx=dev, dtype=dtype) for _, shape in mod.data_shapes]
     batch = mx.io.DataBatch(data, []) # empty label
+    mod.forward(batch, is_train=False)
+    if (dtype == 'float16'):
```
Comment (szha) [inline example/image-classification/benchmark_score.py]: remove parentheses

File: example/image-classification/benchmark_score.py [hunk 3]
```diff
@@ -39,21 +39,32 @@ def get_symbol(network, batch_size):
...
+        arg_params, aux_params = mod.get_params()
+        for key in arg_params.keys():
+            val = arg_params[key]
+            arg_params[key] = mx.nd.Cast(val, dtype='float16')
```
Comment (szha) [inline example/image-classification/benchmark_score.py]: use dtype variable?

File: example/image-classification/benchmark_score.py [hunk 4]
```diff
@@ -39,21 +39,32 @@ def get_symbol(network, batch_size):
...
+        for key in aux_params.keys():
+            val = aux_params[key]
+            aux_params[key] = mx.nd.Cast(val, dtype='float16')
```
Comment (szha) [inline example/image-classification/benchmark_score.py]: same

File: example/image-classification/benchmark_score.py [hunk 5]
```diff
@@ -74,11 +85,21 @@ def score(network, dev, batch_size, num_batches):
...
-        logging.info('network: %s', net)
-        for d in devs:
-            logging.info('device: %s', d)
-            for b in batch_sizes:
-                speed = score(network=net, dev=d, batch_size=b, num_batches=10)
-                logging.info('batch size %2d, image/sec: %f', b, speed)
+    for dtype in dtypes:
+        for net in networks:
+            # AlexNet doesn't support float16
+            if (net == 'alexnet' and dtype == 'float16'):
```
Comment (szha) [inline example/image-classification/benchmark_score.py]: remove parentheses

File: example/image-classification/benchmark_score.py [hunk 6]
```diff
@@ -74,11 +85,21 @@ def score(network, dev, batch_size, num_batches):
...
+                print("alexnet doesn't support float16")
+                continue
+            logging.info('network: %s', net)
+            for d in devs:
+                if (dtype == 'float16' and d == mx.cpu()):
```
Comment (szha) [inline example/image-classification/benchmark_score.py]: remove parentheses

File: example/image-classification/benchmark_score.py (hunk 2 above)
Comment (szha) [reply example/image-classification/benchmark_score.py]: Also, seems that when binding, the parameters will automatically take the first input's data type. https://github.com/apache/incubator-mxnet/blob/master/python/mxnet/module/module.py#L443

**Pull request closed without merging**